pkg/reconciler/revision/background.go

```diff
@@ -0,0 +1,260 @@
+/*
+Copyright 2020 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package revision
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/google/go-containerregistry/pkg/authn/k8schain"
+	"go.uber.org/zap"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/client-go/util/workqueue"
+	v1 "knative.dev/serving/pkg/apis/serving/v1"
+)
+
+// imageResolver is an interface used mostly to mock digestResolver for tests.
+type imageResolver interface {
+	Resolve(ctx context.Context, image string, opt k8schain.Options, registriesToSkip sets.String) (string, error)
+}
+
+// backgroundResolver performs background downloads of image digests.
+type backgroundResolver struct {
+	logger *zap.SugaredLogger
+
+	resolver imageResolver
+	enqueue  func(types.NamespacedName)
+
+	queue workqueue.Interface
+
+	mu      sync.Mutex
+	results map[types.NamespacedName]*resolveResult
+}
+
+// resolveResult is the overall result for a particular revision. We create a
+// workItem for each container we need to resolve for the overall result.
+type resolveResult struct {
+	// these fields are immutable afer creation, so can be accessed without a lock.
+	opt                k8schain.Options
+	registriesToSkip   sets.String
+	completionCallback func()
+
+	// these fields can be written concurrently, so should only be accessed while
+	// holding the backgroundResolver mutex.
+	statuses  []v1.ContainerStatus
+	err       error
+	remaining int
+}
+
+// workItem is a single task submitted to the queue, to resolve a single image
+// for a resolveResult.
+type workItem struct {
+	result  *resolveResult
```

Comment:
The workqueue deduplicates by value, so you should check that we get the proper deduplication behavior the way this is.  Maybe a unit test with a transport that blocks?

Reply:
I may (well :)) be missing something, but I don't think we're actually using the workqueue dedupe here, since we check the results map (under mutex) in order to return the existing result before we'd ever add twice to the queue.

Reply:
It's a little weird, but IIRC the prober also holds the workqueue "wrong" in a similar way and seems fine, so 🤷